### PR TITLE
Add a docker-compose stack

### PR DESCRIPTION
**Pull request by lunika**
### Purpose

To ease the development of this project we added a docker-compose stack.
We need postgresql, redis and nextcloud containers.
A makefile is also added

### Proposal

- [x] Create a `docker-compose` configuration
- [x] Add a `Makefile`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	@$(COMPOSE_RUN) dockerize -wait tcp://db:5432 -timeout 60s
+	@$(COMPOSE_RUN) --entrypoint install.sh nextcloud-install
+.phony: install
+
+run: ## Start the development nextcloud server using Docker
+	@$(COMPOSE) up -d db
+	@echo "Wait for postgresql to be up..."
+	@$(COMPOSE_RUN) dockerize -wait tcp://db:5432 -timeout 60s
+	@$(COMPOSE) up -d nextcloud
+
```
Comment (jmaupetit) [inline Makefile]: Extra empty row.

File: env.d/nextcloud [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+NEXTCLOUD_REDIS_HOST=redis
+NEXTCLOUD_REDIS_PORT=6379
+POSTGRES_DB=nextcloud
+POSTGRES_HOST=db
+POSTGRES_PORT=5432
+POSTGRES_PASSWORD=pass
+POSTGRES_USER=nextcloud_user
+NEXTCLOUD_PASSWORD_SALT=salt
+NEXTCLOUD_SECRET=secret
+INSTANCE_ID=foobarbaz
```
Comment (jmaupetit) [inline env.d/nextcloud]: ab sorting?

File: env.d/db [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# Database
+POSTGRES_DB=nextcloud
+POSTGRES_USER=nextcloud_user
+POSTGRES_PASSWORD=pass
```
Comment (jmaupetit) [inline env.d/db:3]: ab sorting?

File: env.d/sync.dist [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+POSTGRES_PORT=5432
+POSTGRES_PASSWORD=pass
+POSTGRES_USER=nextcloud_user
+NEXTCLOUD_IDPS_METADATA_URL=https://metadata-endpoint.xml
+SAML_DISPLAYNAME_MAPPING=urn:oid:2.5.4.3
+SAML_EMAIL_MAPPING=urn:oid:0.9.2342.19200300.100.1.3
+SAML_GROUP_MAPPING=q
+SAM_GENERAL_UID_MAPPING=urn:oid:1.3.6.1.4.1.5923.1.1.1.6
+SAML_SERVICE_PROVIDER_X509_CERT=/private/service_provider/x509.cert
+SAML_SERVICE_PROVIDER_PK=/private/service_provider/private_key
```
Comment (jmaupetit) [inline env.d/sync.dist:6]: ab sorting?

**Comment by jmaupetit** [comment]
Maybe you should add a README in this PR?

**Comment by lunika** [comment]
> Maybe you should add a README in this PR?

Done my master.

File: README.md [hunk 5]
```diff
@@ -1 +1,64 @@
 # nextcloud-shibboleth-sync
+
+This repo contains an app to synchronize a list of identity providers with the
```
Comment (jmaupetit) [inline README.md]:
```suggestion
This repository contains an app to synchronize a list of identity providers with the 
```

File: README.md [hunk 6]
```diff
@@ -1 +1,64 @@
 # nextcloud-shibboleth-sync
+
+This repo contains an app to synchronize a list of identity providers with the 
+[user_saml](https://github.com/nextcloud/user_saml) Nextcloud application.
+
+This app is written in pure php without dependency and connect to a postgresql database.
```
Comment (jmaupetit) [inline README.md]:
```suggestion
This app is written in pure PHP with no other dependency; it supposes that your NextCloud instance is connected to a PostgreSQL database.
```

File: README.md [hunk 7]
```diff
@@ -1 +1,64 @@
...
+
+## How to build the image
+
+The docker image will copy the `app` directory and the `CMD` instruction directly executes the script.
```
Comment (jmaupetit) [inline README.md:11]: You should maybe say that the base image is our home made one? And add a link to it?

File: README.md [hunk 8]
```diff
@@ -1 +1,64 @@
...
+
+To build it, run:
+
+```bash
+$ docker build -t fundocker/nextcloud-shibboleth-sync:{TAG} .
```
Comment (jmaupetit) [inline README.md]: `TAG` is an example tag or an environment variable?

File: README.md [hunk 9]
```diff
@@ -1 +1,64 @@
...
+- `SAML_SERVICE_PROVIDER_PK`: path to the service provider private key
+- `SAML_SERVICE_PROVIDER_X509_CERT`: path to the service provider X509 certificate
+- `POSTGRES_DB`: postgresql database name
+- `POSTGRES_HOST`: postgresql database host
+- `POSTGRES_PASSWORD`: postgresql database password
+- `POSTGRES_PORT`: postgresql database port
+- `POSTGRES_USER`: postgresql database username
+
+A volume is also used to add the X509 certificate and the private key provided by the identity provider,
+this volume must be mounted in the `/private` directory and the path to each file added to the environment
```
Comment (jmaupetit) [inline README.md]: Relative path to `/private` or absolute path?

File: README.md [hunk 10]
```diff
@@ -1 +1,64 @@
...
+- `POSTGRES_HOST`: postgresql database host
+- `POSTGRES_PASSWORD`: postgresql database password
+- `POSTGRES_PORT`: postgresql database port
+- `POSTGRES_USER`: postgresql database username
+
+A volume is also used to add the X509 certificate and the private key provided by the identity provider,
+this volume must be mounted in the `/private` directory and the path to each file added to the environment 
+variables `SAML_SERVICE_PROVIDER_PK` and `SAML_SERVICE_PROVIDER_X509_CERT`.
+
+A complete exemple is available in the `docker-compose.yml` file.
```
Comment (jmaupetit) [inline README.md]:
```suggestion
A complete example is available in the [`docker-compose.yml`](./docker-compose.yml) file.
```

File: README.md (hunk 8 above)
Comment (jmaupetit) [reply README.md]: Also: why not using `make build`?

File: README.md [hunk 11]
```diff
@@ -1 +1,64 @@
...
+
+## Development workflow
+
+To ease the development, a docker-compose configuration is added. To use this docker-compose environment
+a `Makefile` is here to execute all needed operations.
+
+First you need to bootstrap the project
```
Comment (jmaupetit) [inline README.md]:
```suggestion
First you need to bootstrap the project:
```

File: README.md [hunk 12]
```diff
@@ -1 +1,64 @@
...
+
+```bash
+$ make bootstrap
+```
+
+This task will build all needed images and then install Nextcloud.
+
+Once the project bootstrap you can run the `sync` task
```
Comment (jmaupetit) [inline README.md]:
```suggestion
Once the project has been bootstrapped you can run the `sync` task
```

File: README.md [hunk 13]
```diff
@@ -1 +1,64 @@
...
+
+```
+$ make sync
+```
+
+If you want to run Nextcloud, you have to execute the `run` task and Nextcloud will be available at address: http://localhost:9000
```
Comment (jmaupetit) [inline README.md]:
```suggestion
If you want to run Nextcloud, you have to execute the `run` task and Nextcloud will be available at the following address: http://localhost:9000
```

**Review by jmaupetit — COMMENTED** [review]
Thank you for this @lunika!

File: README.md (hunk 7 above)
Comment (lunika) [reply README.md:11]: No it's the `php` image.

File: README.md (hunk 8 above)
Comment (lunika) [reply README.md]: Because the `make build` build also the `nextcloud`. It is here only to drive the development task

File: README.md (hunk 9 above)
Comment (lunika) [reply README.md]: the absolute path `/private`

File: README.md (hunk 8 above)
Comment (jmaupetit) [reply README.md]: Maybe this Makefile target should be split into two sub-tasks? You did not answer my first question about what `{TAG}` means to you.

File: README.md (hunk 9 above)
Comment (jmaupetit) [reply README.md]: So maybe it's worth to be mentioned explicitly?

Comment (lunika) [reply README.md]: starting a path with a `/` is not explicit ? I though it was a unix-like convention

File: README.md (hunk 8 above)
Comment (lunika) [reply README.md]: sorry, `{TAG}` should be replaced with a real tag, I choose a template like syntax, maybe you have something better to suggest ?

Comment (jmaupetit) [reply README.md]: the thing is that having `{}` suggests a variable to me, why not something like:

```suggestion
$ docker build -t fundocker/nextcloud-shibboleth-sync:my_tag .
```

File: README.md (hunk 9 above)
Comment (jmaupetit) [reply README.md]: I am talking about the values of the two environment variables not the volume. Your expression may suggest that environment variable values are relative to the volume mount path.

File: README.md (hunk 8 above)
Comment (lunika) [reply README.md]: without `$` ?

Comment (lunika) [reply README.md]: I made the change.

Comment (jmaupetit) [reply README.md]: This was confusing to me :shrug:

File: README.md (hunk 9 above)
Comment (lunika) [reply README.md]: ok done

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**